Introduce `QueryEnhancerSelector` to configure which `QueryEnhancerFactory` to use

mp911de opened the pull request:
This is a draft for configuring a `QueryEnhancerSelector` that selects a `QueryEnhancerFactory` based on a `DeclaredQuery`:

```java
@EnableJpaRepositories(queryEnhancerSelector = MyQueryEnhancerSelector.class)

class MyQueryEnhancerSelector extends QueryEnhancerSelector.DefaultQueryEnhancerSelector {
	public MyQueryEnhancerSelector() {
		super(QueryEnhancerFactories.fallback(), DefaultQueryEnhancerSelector.jpql());
	}
}
```

This change requires decoupling of `DeclaredQuery` (the declaration aspect) and the introspected part (`IntrospectedQuery`) to avoid the logical cycle of introspecting a query upon `DeclaredQuery` creation.

Closes #3622

schauder commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/support/JpaRepositoryFactoryBean.java:
```diff
@@ -45,10 +51,12 @@
 	private EntityPathResolver entityPathResolver;
 	private EscapeCharacter escapeCharacter = EscapeCharacter.DEFAULT;
 	private JpaQueryMethodFactory queryMethodFactory;
+	private @Nullable Function<BeanFactory, QueryEnhancerSelector> queryEnhancerSelector;
```
The name is misleading. `queryEnhancerSelectorSource` of course we could go with `queryEnhancerSelectorSelector` ;-)

schauder commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/support/JpaRepositoryFactoryBean.java:
```diff
@@ -100,6 +114,42 @@ public void setQueryMethodFactory(@Nullable JpaQueryMethodFactory factory) {
+	 *
+	 * @param queryEnhancerSelector must not be {@literal null}.
+	 */
+	public void setQueryEnhancerSelector(Class<? extends QueryEnhancerSelector> queryEnhancerSelector) {
```
`queryEnhancerSelectorType`?

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/DefaultDeclaredQuery.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * @author Mark Paluch
+ */
+class DefaultDeclaredQuery implements DeclaredQuery {
```
I deliberately do not want to use records to avoid performance penalties on GraalVM. Once these are sorted out, we can switch to records.

mp911de commented on spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/JpaQueryConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @author Mark Paluch
+ */
+public class JpaQueryConfiguration {
```
I do not like records being public API.

mp911de commented:
I addressed a few review comments and introduced `EntityQuery` as additional interface. Likely, there's a better name since queries can return projections or aggregations when using native SQL. In any case, we signal that only `EntityQuery` allows count query derivation.

I think this change still requires some refinement before I can turn it into a PR.

mp911de commented:
That's merged now.
